File: CourseraAlgorithmicToolbox Problems/week3_greedy_algorithms/5_collecting_signatures/CoveringSegments.kt
import java.util.*

fun optimalPoints(segments: MutableList<Segment>): IntArray {
    //write your code here
    val points = IntArray(2 * segments.size)
    for (i in segments.indices) {
        points[2 * i] = segments[i].start
        points[2 * i + 1] = segments[i].end
    }
    return points
}

class Segment(var start: Int, var end: Int)

fun main(args: Array<String>) {
    val scanner = Scanner(System.`in`)
    val n = scanner.nextInt()
    val segments = ArrayList<Segment>(n)
    repeat (n) {
        val start: Int = scanner.nextInt()
        val end: Int = scanner.nextInt()
        segments += Segment(start, end)
    }
    val points = optimalPoints(segments)
    println(points.size)
    for (point in points) {
        print(point.toString() + " ")
    }
}

File: Resources/Coursera/AlgorithmicToolbox/week3_greedy_algorithms/5_collecting_signatures/by_learners/covering_segments.rb
#!/usr/bin/env ruby
# by Andronik Ordian

Segment = Struct.new("Segment", :start, :end)

def optimal_points(segments)
  points = []
  return points if segments.empty?
  #write your code here
  points
end

if __FILE__ == $0
  data = STDIN.read.split().map(&:to_i)
  n = data[0]
  segments = data[1..2*n].each_slice(2).to_a.map { |e| Segment.new(e[0], e[1]) }
  points = optimal_points(segments)
  puts "#{points.size}"
  puts "#{points.join(' ')}"
end